File: 1671-minimum-number-of-removals-to-make-mountain-array/1671-minimum-number-of-removals-to-make-mountain-array.c
#include <stdio.h>
#include <stdlib.h>

// Define MAX macro for easy readability
#define MAX(a, b) ((a) > (b) ? (a) : (b))

/**
 * @brief Binary search helper to find the insertion point in a sorted array.
 *
 * @param arr (int*): Array in which to search.
 * @param size (int): Current size of the array.
 * @param target (int): Target value to search.
 * @return (int): Position where target should be placed.
 */
int binarySearchInsertPosition(int* arr, int size, int target) {
    int left = 0, right = size;
    while (left < right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid;
    }
    return left;
}

/**
 * @brief Calculates the minimum number of elements to remove to make the array a mountain array.
 *
 * @param nums (int*): Pointer to the input array of integers.
 * @param numsSize (int): Size of the input array.
 * @return (int): Minimum number of elements to remove to form a mountain array.
 */
int minimumMountainRemovals(int* nums, int numsSize) {
    int* lis_length = (int*)malloc(numsSize * sizeof(int));
    int* lds_length = (int*)malloc(numsSize * sizeof(int));
    int* aux_seq = (int*)malloc(numsSize * sizeof(int));
    
    int seq_len, max_mountain_length = 0;

    // Step 1: Compute longest increasing subsequence (LIS) lengths ending at each index
    seq_len = 0;
    for (int i = 0; i < numsSize; i++) {
        int pos = binarySearchInsertPosition(aux_seq, seq_len, nums[i]);
        if (pos == seq_len) {
            aux_seq[seq_len++] = nums[i];
        } else {
            aux_seq[pos] = nums[i];
        }
        lis_length[i] = pos + 1;
    }

    // Step 2: Compute longest decreasing subsequence (LDS) lengths starting at each index
    seq_len = 0;
    for (int i = numsSize - 1; i >= 0; i--) {
        int pos = binarySearchInsertPosition(aux_seq, seq_len, nums[i]);
        if (pos == seq_len) {
            aux_seq[seq_len++] = nums[i];
        } else {
            aux_seq[pos] = nums[i];
        }
        lds_length[i] = pos + 1;
    }

    // Step 3: Find the maximum length of a valid mountain subarray
    for (int i = 1; i < numsSize - 1; i++) {
        if (lis_length[i] > 1 && lds_length[i] > 1) {
            int mountain_length = lis_length[i] + lds_length[i] - 1;
            max_mountain_length = MAX(max_mountain_length, mountain_length);
        }
    }

    // Step 4: Calculate minimum removals and free memory
    int min_removals = numsSize - max_mountain_length;
    
    free(lis_length);
    free(lds_length);
    free(aux_seq);
    
    return min_removals;
}

File: 1671-minimum-number-of-removals-to-make-mountain-array/1671-minimum-number-of-removals-to-make-mountain-array.py
from bisect import bisect_left

class Solution:
    def minimumMountainRemovals(self, nums: list[int]) -> int:
        """
        Calculate the minimum number of elements to remove to make the array a mountain array.

        Parameters:
        nums (list[int]): List of integers representing the input array.

        Returns:
        int: Minimum number of elements to remove to form a mountain array.
        """
        n = len(nums)
        
        # Step 1: Compute longest increasing subsequence (LIS) ending at each index
        lis_length = [0] * n  # lis_length[i] will store the LIS ending at index i
        lis_seq = []          # Auxiliary list for binary search based LIS
        
        for i in range(n):
            pos = bisect_left(lis_seq, nums[i])
            if pos == len(lis_seq):
                lis_seq.append(nums[i])  # Extend the sequence
            else:
                lis_seq[pos] = nums[i]   # Replace the element at pos
            lis_length[i] = pos + 1      # Length of LIS ending at index i

        # Step 2: Compute longest decreasing subsequence (LDS) starting at each index
        lds_length = [0] * n  # lds_length[i] will store the LDS starting at index i
        lds_seq = []          # Auxiliary list for binary search based LDS
        
        for i in range(n - 1, -1, -1):
            pos = bisect_left(lds_seq, nums[i])
            if pos == len(lds_seq):
                lds_seq.append(nums[i])  # Extend the sequence
            else:
                lds_seq[pos] = nums[i]   # Replace the element at pos
            lds_length[i] = pos + 1      # Length of LDS starting at index i

        # Step 3: Find the maximum length of a mountain array
        max_mountain_length = 0
        for i in range(1, n - 1):
            # A valid peak must have both an increasing sequence on the left and a decreasing on the right
            if lis_length[i] > 1 and lds_length[i] > 1:
                # Calculate mountain length centered at i and update max length
                mountain_length = lis_length[i] + lds_length[i] - 1
                max_mountain_length = max(max_mountain_length, mountain_length)

        # Step 4: Calculate minimum removals
        return n - max_mountain_length
